contracts/signature_mint/SignatureMint.sol

```diff
@@ -0,0 +1,276 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+// Interface
+import { ISignatureMint } from "./ISignatureMint.sol";
+
+// Token
+import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
+
+// Protocol control center.
+import { ProtocolControl } from "../ProtocolControl.sol";
+
+// Signature utils
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "@openzeppelin/contracts/utils/cryptography/draft-EIP712.sol";
+
+// Access Control + security
+import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+
+// Royalties
+import "@openzeppelin/contracts/interfaces/IERC2981.sol";
+
+// Meta transactions
+import "@openzeppelin/contracts/metatx/ERC2771Context.sol";
+
+// Utils
+import "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/utils/Multicall.sol";
+
+contract SignatureMint is
+    
+    ISignatureMint,
+    ERC721Enumerable,
+    EIP712,
+    AccessControlEnumerable,
+    ERC2771Context,
+    IERC2981,
+    ReentrancyGuard,
+    Multicall
+
+{
+    using ECDSA for bytes32;
+    using Strings for uint256;
+
+    bytes32 private constant TYPEHASH =
+        keccak256("MintRequest(address to,string baseURI,uint256 amountToMint,uint256 validityStartTimestamp,uint256 validityEndTimestamp,bytes uid)");
+
+    /// @dev Only TRANSFER_ROLE holders can have tokens transferred from or to them, during restricted transfers.
+    bytes32 public constant TRANSFER_ROLE = keccak256("TRANSFER_ROLE");
+    /// @dev Only MINTER_ROLE holders can sign off on `MintRequest`s.
+    bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+
+    /// @dev The token ID of the next token to mint.
+    uint256 public nextTokenIdToMint;
+
+    /// @dev Contract interprets 10_000 as 100%.
+    uint64 private constant MAX_BPS = 10_000;
+
+    /// @dev The % of secondary sales collected as royalties. See EIP 2981.
+    uint64 public royaltyBps;
+
+    /// @dev The % of primary sales collected by the contract as fees.
+    uint120 public feeBps;
+
+    /// @dev Whether transfers on tokens are restricted.
+    bool public transfersRestricted;
+
+    /// @dev Contract level metadata.
+    string public contractURI;
+    
+    /// @dev The protocol control center.
+    ProtocolControl internal controlCenter;
+
+    uint256[] private baseURIIndices;
+    
+    /// @dev Mapping from end-tokenId => baseURI.
+    mapping(uint256 => string) public baseURI;
+
+    /// @dev Mapping from mint request UID => whether the mint request is processed.
+    mapping(bytes => bool) private minted;
```

Comment:
could we just use tokenId to keep track of this?
signature includes tokenId to mint and we can do owner(tokenId) to verify validity. 

or what's the reason behind creating a request UID?

Reply:
The reason for using a `uid` is to prevent replay attacks i.e. calling `mint` with the same mint request and signature, multiple times before the mint request expires.

The reason we don't use `tokenId` in the `minted` mapping:
- Tokens are minted with an incrementing `nextTokenIdToMint` like in our other NFT contracts.
- Logic in the contract depends on this type of incrementing. See [`assignURI`](https://github.com/nftlabs/nftlabs-protocols/blob/sig-based-minting/contracts/signature_mint/SignatureMint.sol#L215-L225) and our general gas efficient pattern for assigning the same base URI to batches of token IDs.

The reason we don't accept `tokenIds[]` in a mint request:

The UI will need to stay in sync with the contract's state. Say, there are 2 end users for whom a mint request is generated at nearly the same time. If we accept token IDs in the mint request, the token IDs in those two different mint requests will clash, resulting in one of those mint requests to revert, even though the signature and all other parameters are valid.